File: csharp/src/CommandPattern/MoveForwardCommand.cs
using System;
using System.Collections.Generic;

namespace CommandPattern
{
    public class MoveForwardCommand : ICommand
    {
        private Robot _robot;
        private IDictionary<string, Action<Robot>> _moveActions;

        public MoveForwardCommand(Robot robot)
        {
            _robot = robot;
            _moveActions = new Dictionary<string, Action<Robot>>
                               {
                                   {"N", r => r.Position.Y += 1 },
                                   {"E", r => r.Position.X += 1 },
                                   {"S", r => r.Position.Y -= 1 },
                                   {"W", r => r.Position.X -= 1 }
                               };
        }
 
        public void Execute()
        {
            _moveActions[_robot.Direction](_robot);
        }

        public void Undo()
        {
            var oppositeDirection = Compass.GetOppositeDirection(_robot.Direction);
            _moveActions[oppositeDirection](_robot);
        }
    }
}

File: ruby/lib/move_forward_command.rb
class MoveForwardCommand
	def initialize(robot)
		@robot = robot
		@move_actions = {:N => Proc.new{|r| r.position.y += 1},
						:E => Proc.new{|r| r.position.x += 1},
						:S => Proc.new{|r| r.position.y -= 1},
						:W => Proc.new{|r| r.position.x -= 1}}
	end

	def execute
	  @move_actions[@robot.direction].call @robot
    end

    def undo
      opposite_direction = Compass.get_opposite_direction @robot.direction
      @move_actions[opposite_direction].call @robot
    end
end